File: android/app/src/main/java/app/headstart/data/SpotRepository.kt
package app.headstart.data

import app.headstart.core.Callables
import app.headstart.core.HeadstartError
import app.headstart.core.snapshotFlow
import app.headstart.core.str
import com.google.firebase.firestore.FirebaseFirestore
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map

/** Mirrors the clamps in `upsertSpot` (addendum §K) so the UI can never propose a rejected value. */
object SpotLimits {
    const val MIN_LEAD_MIN = 1
    const val MAX_LEAD_MIN = 30
    const val DEFAULT_LEAD_MIN = 3

    /** SpotEdit.dc.html shows a 1–15 slider; the contract allows up to 30. */
    const val SLIDER_MAX_LEAD_MIN = 15

    const val MIN_RADIUS_M = 50
    const val MAX_RADIUS_M = 500
    const val DEFAULT_RADIUS_M = 100
    const val MAX_NAME_LENGTH = 40

    val RADIUS_PRESETS = listOf(50, 100, 200, 300, 500)

    fun clampLeadTimeMin(value: Int): Int = value.coerceIn(MIN_LEAD_MIN, MAX_LEAD_MIN)

    fun clampRadiusM(value: Int): Int = value.coerceIn(MIN_RADIUS_M, MAX_RADIUS_M)

    fun nextRadius(current: Int): Int {
        val at = RADIUS_PRESETS.indexOf(current)
        return if (at < 0) RADIUS_PRESETS.first() else RADIUS_PRESETS[(at + 1) % RADIUS_PRESETS.size]
    }

    fun validName(raw: String): String? {
        val trimmed = raw.trim()
        return if (trimmed.isEmpty() || trimmed.length > MAX_NAME_LENGTH) null else trimmed
    }

    fun validCoords(lat: Double, lng: Double): Boolean =
        !lat.isNaN() && !lng.isNaN() && lat in -90.0..90.0 && lng in -180.0..180.0
}

class SpotRepository(
    private val db: FirebaseFirestore,
    private val callables: Callables,
) {
    /** Live list for the pair. Single-field equality — no composite index needed. */
    fun spots(pairId: String): Flow<List<Spot>> =
        db.collection("spots")
            .whereEqualTo("pairId", pairId)
            .snapshotFlow()
            .map { snap ->
                snap.documents.mapNotNull { spotFrom(it.id, it.data) }.sortedBy { it.name.lowercase() }
            }

    /** Creates when [spotId] is null, updates otherwise. Returns the spot id. */
    suspend fun upsertSpot(
        pairId: String,
        name: String,
        lat: Double,
        lng: Double,
        leadTimeMin: Int,
        radiusM: Int,
        spotId: String? = null,
    ): String {
        val cleanName = SpotLimits.validName(name) ?: throw HeadstartError.BadName
        if (!SpotLimits.validCoords(lat, lng)) throw HeadstartError.BadCoords
        val payload = mutableMapOf<String, Any?>(
            "pairId" to pairId,
            "name" to cleanName,
            "lat" to lat,
            "lng" to lng,
            "leadTimeMin" to SpotLimits.clampLeadTimeMin(leadTimeMin),
            "radiusM" to SpotLimits.clampRadiusM(radiusM),
        )
        if (spotId != null) payload["spotId"] = spotId
        val res = callables.call("upsertSpot", payload)
        return res.str("spotId") ?: spotId ?: throw HeadstartError.Unknown("upsertSpot-no-id")
    }

    suspend fun deleteSpot(spotId: String) {
        callables.call("deleteSpot", mapOf("spotId" to spotId))
    }
}

File: ios/Headstart/Data/SpotRepository.swift
// ios/Headstart/Data/SpotRepository.swift
//
// `spots where pairId == …` is a single-field query, so no composite index is needed.
//
// ADDENDUM §K — this repository is where the client-side clamps live. The server clamps
// `leadTimeMin` to 1–30 and `radiusM` to 50–500 too, but an out-of-range value that reaches
// the wire comes back as a callable error a person cannot interpret. Clamping here AND
// constraining the editor's controls means an illegal value is never expressible in the
// first place. `Callables.upsertSpot` clamps a third time at the wire boundary; the clamps
// are idempotent, so the belt and the braces cost nothing.

import Foundation
import FirebaseFirestore

@MainActor
public final class SpotRepository: ObservableObject {

    @Published public private(set) var spots: [Spot] = []
    @Published public private(set) var hasLoaded = false

    private let db: Firestore
    private let callables: Callables
    private var task: Task<Void, Never>?

    public init(db: Firestore, callables: Callables) {
        self.db = db
        self.callables = callables
    }

    public func observe(pairId: String) {
        task?.cancel()
        hasLoaded = false
        let stream = db.collection("spots")
            .whereField("pairId", isEqualTo: pairId)
            .modelsStream { Spot(id: $0, data: $1) }
        task = Task { [weak self] in
            for await spots in stream {
                guard let self else { return }
                // Oldest first, so the list does not reshuffle under a thumb when the
                // server backfills a field on an existing spot.
                self.spots = spots.sorted { $0.createdAt < $1.createdAt }
                self.hasLoaded = true
            }
        }
    }

    public func stop() {
        task?.cancel()
        task = nil
        spots = []
        hasLoaded = false
    }

    public func spot(id: String) -> Spot? { spots.first { $0.id == id } }

    // MARK: - Callables

    /// Errors: `.badName` (empty or >40 characters), `.badCoords` (non-finite), `.notPaired`.
    /// `leadTimeMin` and `radiusM` are clamped to the contract's ranges before the call.
    @discardableResult
    public func upsert(
        pairId: String,
        name: String,
        lat: Double,
        lng: Double,
        leadTimeMin: Int,
        radiusM: Double,
        spotId: String? = nil
    ) async throws -> String {
        let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { throw HeadstartError.badName }
        guard lat.isFinite, lng.isFinite,
              abs(lat) <= 90, abs(lng) <= 180 else { throw HeadstartError.badCoords }
        // ADDENDUM §K: leadTimeMin 1–30, radiusM 50–500.
        let clampedLeadTimeMin = SpotLimits.clampLeadTimeMin(leadTimeMin)
        let clampedRadiusM = SpotLimits.clampRadiusM(radiusM)
        let response = try await callables.upsertSpot(
            pairId: pairId,
            name: String(trimmed.prefix(40)),
            lat: lat,
            lng: lng,
            leadTimeMin: clampedLeadTimeMin,
            radiusM: clampedRadiusM,
            spotId: spotId
        )
        return response.spotId
    }

    public func delete(spotId: String) async throws {
        try await callables.deleteSpot(spotId: spotId)
    }
}
